### PR TITLE
fix: Use Pinata ipfs instead of web3.storage

### LINKED ISSUE #358: Update the the IPFS upload task to use the new web3.storage api
Web3.storage has updated their API and account system.

As a result, the integration needs to be updated, as the current api will be sunset on 2024-01-09

Announcement here: https://blog.web3.storage/posts/the-data-layer-is-here-with-the-new-web3-storage

### PR DESCRIPTION
**Pull request by negar-abbasi**
## Proposed Changes

  - Using Piñata instead of web3.storage due to changes in web3.storage
  - Add yaspin package for showing the uploading bar

closes #358

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ auth0-python = "^4.4.0"
 algokit-utils = "v2.1.0"
 multiformats = "^0.2.1"
 aiohttp = "3.9.1"
+yaspin = "^3.0.1"
```
Comment (robdmoore) [inline pyproject.toml:27]: Is there any web3 storage packages we need to remove?

**Review by robdmoore — APPROVED** [review]
Nice work!

File: pyproject.toml (hunk 1 above)
Comment (negar-abbasi) [reply pyproject.toml:27]: There doesn't seem to be anything to remove.

File: src/algokit/core/tasks/ipfs.py [hunk 2]
```diff
@@ -1,111 +1,118 @@
...
+        old_api_key = keyring.get_password("algokit_web3_storage", "algokit_web3_storage_access_token")
+        if old_api_key:
+            logger.warning(
+                "You are using the old Web3 Storage API key. Please login again using `algokit ipfs login` with Pinata "
+                "ipfs provider. Follow the instructions on https://docs.pinata.cloud/docs/getting-started "
+                "to create an account and obtain a JWT."
+            )
+            keyring.delete_password("algokit_web3_storage", "algokit_web3_storage_access_token")
+    except KeyError as ex:
+        logger.debug("Old Web3 Storage API key is already deleted.", exc_info=ex)
```
Comment (neilcampbell) [inline src/algokit/core/tasks/ipfs.py:67]: This exception will be thrown and caught under normal conditions, so I feel we shouldn't have a debug message here.

File: src/algokit/core/tasks/ipfs.py [hunk 3]
```diff
@@ -1,111 +1,118 @@
...
+    return keyring.get_password(ALGOKIT_PINATA_NAMESPACE, ALGOKIT_PINATA_TOKEN_KEY)
+
+
+def set_pinata_jwt(jwt: str | None) -> None:
     """
-    Sets or deletes a password in the keyring library based on the provided API key.
+    Sets or deletes a password in the keyring library based on the provided KWT.
```
Comment (neilcampbell) [inline src/algokit/core/tasks/ipfs.py:73]:
```suggestion
    Sets or deletes a password in the keyring library based on the provided JWT.
```

File: src/algokit/core/tasks/ipfs.py [hunk 4]
```diff
@@ -115,31 +122,34 @@ def upload_to_web3_storage(
...
     try:
         response = httpx.post(
-            "https://api.web3.storage/upload",
+            url="https://api.pinata.cloud/pinning/pinFileToIPFS",
+            data=data,
+            files=files,
             headers=headers,
-            timeout=timeout,
-            content=content or file_content,
+            timeout=DEFAULT_TIMEOUT,
```
Comment (neilcampbell) [inline src/algokit/core/tasks/ipfs.py:138]: Looks like the `DEFAULT_TIMEOUT` is 10s. We may want to bump that up to account for larger files. Maybe 90s?

File: src/algokit/core/tasks/ipfs.py [hunk 5]
```diff
@@ -115,31 +122,34 @@ def upload_to_web3_storage(
...
         )
         response.raise_for_status()
-        cid = response.json().get("cid")
-        if not isinstance(cid, str):
-            raise ValueError("CID is not a string.")
-        return cid
+        ipfs_hash = response.json().get("IpfsHash")
```
Comment (neilcampbell) [inline src/algokit/core/tasks/ipfs.py:141]: You can keep the variable named cid if you like. Cid is a hash of the file contents.

File: src/algokit/core/tasks/mint/mint.py [hunk 6]
```diff
@@ -159,7 +159,7 @@ def mint_token(  # noqa: PLR0913
 
     Args:
         client (algod.AlgodClient): An instance of the `algod.AlgodClient` class representing the Algorand node.
-        api_key (str): A string representing the API key for accessing the Algorand network.
+        jwt (str): A string representing the JWT for accessing the Algorand network.
```
Comment (neilcampbell) [inline src/algokit/core/tasks/mint/mint.py:162]:
```suggestion
        jwt (str): The JWT for accessing the Piñata API.
```

**Review by neilcampbell — APPROVED** [review]

**Pull request merged**